### PR TITLE
Added workflow Executors for Approval workflow executor

### PR DESCRIPTION
**Pull request by SahanHerath**
Added two rest APIs
Added workflow executors for application creation, application subscription, API state change, application registration key generation, and user self sign up

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 1]
```diff
@@ -5977,10 +5981,23 @@ public void addWorkflowEntry(WorkflowDTO workflow) throws APIManagementException
             prepStmt.setString(7, workflow.getTenantDomain());
             prepStmt.setString(8, workflow.getExternalWorkflowReference());
 
-            prepStmt.execute();
+            if(workflow.getMetadata() != null) {
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:5984]: Please format all your code.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/APIStateChangeApprovalWorkflowExecutor.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+            ApiMgtDAO apiMgtDAO = ApiMgtDAO.getInstance();
+            Workflow workflow = apiMgtDAO.getworkflowReferenceByExternalWorkflowReference(externalWorkflowRef);
+            String apiName = workflow.getMetadata("ApiName");
+            String action = workflow.getMetadata("Action");
+            String providerName = workflow.getMetadata("ApiProvider");
+            String version = workflow.getMetadata("ApiVersion");
+            String invoker = workflow.getMetadata("Invoker");
+            String currentStatus = workflow.getMetadata("CurrentState");
+            int tenantId = workflowDTO.getTenantId();
+            //ApiMgtDAO apiMgtDAO = ApiMgtDAO.getInstance();
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/APIStateChangeApprovalWorkflowExecutor.java:138]: removed unwanted code

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/ApplicationCreationApprovalWorkflowExecutor.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+     * @param workflowDTO
+     */
+    @Override
+    public WorkflowResponse execute(WorkflowDTO workflowDTO) throws WorkflowException {
+        if (log.isDebugEnabled()) {
+            log.debug("Executing Application creation Workflow.");
+        }
+        ApplicationWorkflowDTO appWorkFlowDTO = (ApplicationWorkflowDTO) workflowDTO;
+        Application application = appWorkFlowDTO.getApplication();
+        String callBackURL = appWorkFlowDTO.getCallbackUrl();
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/ApplicationCreationApprovalWorkflowExecutor.java:56]: remove unused variables.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 4]
```diff
@@ -14650,4 +14667,243 @@ public String getUserID(String userName) throws APIManagementException {
...
+        }
+        try {
+            connection = APIMgtDBUtil.getConnection();
+            prepStmt = connection.prepareStatement(sqlQuery);
+
+            if (workflowType != null){
+                prepStmt.setString(1, workflowType);
+                prepStmt.setString( 2, status);
+                prepStmt.setString( 3, tenantDomain);
+            } else {
```
[the remark this answers is not in the document]
Comment (SahanHerath) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:14851]: fixed in f5092a9

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 5]
```diff
@@ -14650,4 +14667,243 @@ public String getUserID(String userName) throws APIManagementException {
...
+    }
+
+    /**
+     * Get the Pending workflow Request using ExternalWorkflowReference
+     * @param externalWorkflowRef
+     * @return workflow pending request
+     * @throws APIManagementException
+     */
+    public Workflow getworkflowReferenceByExternalWorkflowReference(String externalWorkflowRef)throws APIManagementException {
+
```
[the remark this answers is not in the document]
Comment (SahanHerath) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:14775]: fixed in f5092a9

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 6]
```diff
@@ -5977,10 +5981,23 @@ public void addWorkflowEntry(WorkflowDTO workflow) throws APIManagementException
             prepStmt.setString(7, workflow.getTenantDomain());
             prepStmt.setString(8, workflow.getExternalWorkflowReference());
 
-            prepStmt.execute();
+            if(workflow.getMetadata() != null) {
+                byte[] metadataByte = workflow.getMetadata().toJSONString().getBytes("UTF-8");
+                prepStmt.setBlob(9, new ByteArrayInputStream(metadataByte) );
+            } else {
+                prepStmt.setBlob(9, (Blob)null);
+            }
```
[the remark this answers is not in the document]
Comment (SahanHerath) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:5990]: changed it to setNull (parameter index, value)

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 7]
```diff
@@ -14723,4 +14739,246 @@ public String getUserID(String userName) throws APIManagementException {
...
+            connection.setAutoCommit(false);
+            prepStmt = connection.prepareStatement(query);
+            prepStmt.setString(1, workflowExtRef);
+            prepStmt.executeUpdate();
+            connection.commit();
+        } catch (SQLException e) {
+            handleException("Failed to delete the workflow request. ", e);
+        } finally {
+            APIMgtDBUtil.closeAllConnections(prepStmt, connection, null);
+        }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:14764]: Use "try with resources" whenever possible. We can do the same functionality in less amount of code.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 8]
```diff
@@ -14723,4 +14739,246 @@ public String getUserID(String userName) throws APIManagementException {
...
+        Workflow[] workflows = null;
+        String sqlQuery;
+        if (workflowType != null) {
+            sqlQuery = SQLConstants.GET_ALL_WORKFLOW_DETAILS_BY_WORKFLOW_TYPE;
+        } else {
+            sqlQuery = SQLConstants.GET_ALL_WORKFLOW_DETAILS;
+        }
+        try {
+            connection = APIMgtDBUtil.getConnection();
+            prepStmt = connection.prepareStatement(sqlQuery);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:14845]: It's good to use try with resources

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/resources/admin-api.yaml [hunk 9]
```diff
@@ -2521,17 +2521,151 @@ paths:
...
-  ######################################################
+# The "Workflow Collection" resource APIs
+######################################################
+  '/workflows':
+
+#---------------------------------------------------
+# Retrieve a list of all workflow pending processes
+#---------------------------------------------------
+    get:
+      x-scope: apim:api_workflow
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/resources/admin-api.yaml:2532]: Use <verb>_<action> format.
EG:
apim:api_workflow_view

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/resources/admin-api.yaml [hunk 10]
```diff
@@ -2521,17 +2521,151 @@ paths:
...
+  #----------------------------------------------------------------
+  # Retrieve workflow requests according to the external reference
+  #----------------------------------------------------------------
+  '/workflows/{externalWorkflowRef}':
+
+    #-----------------------------------------------------
+    # Retrieve the Details of a workflow pendding process
+    #-----------------------------------------------------
+    get:
+      x-scope: apim:api_workflow
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/resources/admin-api.yaml:2611]: apim:api_workflow_view

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/resources/admin-api.yaml [hunk 11]
```diff
@@ -2521,17 +2521,151 @@ paths:
...
   /workflows/update-workflow-status:
 
-    #-------------------------------------------------------------------
-    # Resume the workflow by approving or rejecting the workflow request
-    #-------------------------------------------------------------------
+#-------------------------------------------------------------------
+# Resume the workflow by approving or rejecting the workflow request
+#-------------------------------------------------------------------
     post:
       x-scope: apim:api_workflow
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin/src/main/resources/admin-api.yaml:2666]: x-scope: apim:api_workflow_update

**Comment by Arshardh** [comment]
Please try to get this PR merged soon

**Comment by chamilaadhi** [comment]
@Arshardh we have to wait till the admin rest API v1.0 gets merged to the master branch to merge this PR.(https://github.com/wso2/carbon-apimgt/pull/8453) 

Since this PR has added new resources to the v0.16, we have to move them to the new v1.0 version as well

@malinthaprasan shall we get the v1.0 admin rest API merged to the master branch?

**Comment by Arshardh** [comment]
If this will take more time, shall we close this for the moment and reopen this when ready?
There seems to be many conflicts as well.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/ApplicationCreationApprovalWorkflowExecutor.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+            log.debug("Executing Application creation Workflow.");
+        }
+        ApplicationWorkflowDTO appWorkFlowDTO = (ApplicationWorkflowDTO) workflowDTO;
+        Application application = appWorkFlowDTO.getApplication();
+        String message = "Approve application " + application.getName() + " creation request from application creator -"
+                + appWorkFlowDTO.getUserName() + " with throttling tier - " + application.getTier();
+        workflowDTO.setWorkflowDescription(message);
+        workflowDTO.setProperties("ApplicationName", application.getName());
+        workflowDTO.setProperties("UserName", appWorkFlowDTO.getUserName());
+        workflowDTO.setProperties("ApplicationTier", application.getTier());
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/workflow/ApplicationCreationApprovalWorkflowExecutor.java:61]: lets use camelcase

**Pull request merged**